File: hadoop-0.20.2/src/core/org/apache/hadoop/security/GroupMappingServiceProvider.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.hadoop.security;

import java.io.IOException;
import java.util.List;

/**
 * An interface for the implementation of a user-to-groups mapping service
 * used by {@link Groups}.
 */
// @InterfaceAudience.Public
// @InterfaceStability.Evolving
public interface GroupMappingServiceProvider {
  
  /**
   * Get all various group memberships of a given user.
   * Returns EMPTY list in case of non-existing user
   * @param user User's name
   * @return group memberships of user
   * @throws IOException
   */
  public List<String> getGroups(String user) throws IOException;
  /**
   * Refresh the cache of groups and user mapping
   * @throws IOException
   */
  public void cacheGroupsRefresh() throws IOException;
  /**
   * Caches the group user information
   * @param groups list of groups to add to cache
   * @throws IOException
   */
  public void cacheGroupsAdd(List<String> groups) throws IOException;
}

File: core/src/main/scala/org/apache/spark/security/GroupMappingServiceProvider.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.spark.security

/**
 * This Spark trait is used for mapping a given userName to a set of groups which it belongs to.
 * This is useful for specifying a common group of admins/developers to provide them admin, modify
 * and/or view access rights. Based on whether access control checks are enabled using
 * spark.acls.enable, every time a user tries to access or modify the application, the
 * SecurityManager gets the corresponding groups a user belongs to from the instance of the groups
 * mapping provider specified by the entry spark.user.groups.mapping.
 */

trait GroupMappingServiceProvider {

  /**
   * Get the groups the user belongs to.
   * @param userName User's Name
   * @return set of groups that the user belongs to. Empty in case of an invalid user.
   */
  def getGroups(userName : String) : Set[String]

}
